lsp-mode.el

```diff
@@ -4501,38 +4501,41 @@ When the heuristic fails to find the prefix start point, return DEFAULT value."
   "List all possible completions in cached ITEMS with their prefixes.
 We can pass LSP-ITEMS, which will be used when there's no cache.
 Also, additional data to attached to each candidate can be passed via PLIST."
-  (let ((filtered-items
-         (if items
-             (->> (let (queries fuz-queries)
-                    (-keep (lambda (cand)
-                             (let* ((start-point (get-text-property 0 'lsp-completion-start-point cand))
-                                    (query (or (plist-get queries start-point)
-                                               (let ((s (buffer-substring-no-properties start-point (point))))
-                                                 (setq queries (plist-put queries start-point s))
-                                                 s)))
-                                    (fuz-query (or (plist-get fuz-queries start-point)
-                                                   (let ((s (lsp--regex-fuzzy query)))
-                                                     (setq fuz-queries
-                                                           (plist-put fuz-queries start-point s))
-                                                     s))))
-                               (when (string-match fuz-query cand)
-                                 (setq cand (copy-sequence cand))
-                                 (put-text-property 0 1 'match-data (match-data) cand)
-                                 (put-text-property 0 1 'completion-score (lsp--fuzzy-score query cand) cand)
-                                 cand)))
-                           items))
-                  (-sort (-on #'> (lambda (o)
-                                    (or (get-text-property 0 'sort-score o)
-                                        (let* ((score (* (or (get-text-property 0 'completion-score o)
-                                                             0.001)
-                                                         (or (get-text-property 0 'lsp-completion-score o)
-                                                             0.001))))
-                                          (put-text-property 0 1 'sort-score score o)
-                                          score)))))
-                  ;; TODO: pass additional function to sort the candidates
-                  (-map (-partial #'get-text-property 0 'lsp-completion-item)))
-           lsp-items)))
-    (-map (apply #'-rpartial #'lsp--make-completion-item plist) filtered-items)))
+  (let ((res (while-no-input
```

Comment:
can we wrap the whole all-completion?

Reply:
I don't think we need and should wrap `all-completion`. Since it can be interrupted in middle of calculating some thing like cache, we may have to live with that broken cache until either user have to abort completion of complete completion with broken candidates.

Just wrap some thing that's expensive and keep the other preparing/cleaning up code intact is better IMO.

If you want to have `while-no-input` to interrupting always, you can try https://github.com/kiennq/company-capf-noblock, which will forcefully abort current completion at point with `while-no-input`

Reply:
We should if can go upstream.

About the expensive parts - I think that the initial sort is slow too. But let's get benchmarks for that.